File: src/main/java/com/freya02/botcommands/internal/application/ApplicationCommandsBuilder.java
package com.freya02.botcommands.internal.application;

import com.freya02.botcommands.api.Logging;
import com.freya02.botcommands.api.application.ApplicationCommand;
import com.freya02.botcommands.api.application.CommandPath;
import com.freya02.botcommands.api.application.CommandUpdateResult;
import com.freya02.botcommands.api.application.context.annotations.JDAMessageCommand;
import com.freya02.botcommands.api.application.context.annotations.JDAUserCommand;
import com.freya02.botcommands.api.application.context.message.GlobalMessageEvent;
import com.freya02.botcommands.api.application.context.message.GuildMessageEvent;
import com.freya02.botcommands.api.application.context.user.GlobalUserEvent;
import com.freya02.botcommands.api.application.context.user.GuildUserEvent;
import com.freya02.botcommands.api.application.slash.GlobalSlashEvent;
import com.freya02.botcommands.api.application.slash.GuildSlashEvent;
import com.freya02.botcommands.api.application.slash.annotations.JDASlashCommand;
import com.freya02.botcommands.internal.BContextImpl;
import com.freya02.botcommands.internal.application.context.message.MessageCommandInfo;
import com.freya02.botcommands.internal.application.context.user.UserCommandInfo;
import com.freya02.botcommands.internal.application.slash.SlashCommandInfo;
import com.freya02.botcommands.internal.utils.ReflectionUtils;
import com.freya02.botcommands.internal.utils.Utils;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.exceptions.ErrorResponseException;
import net.dv8tion.jda.api.requests.ErrorResponse;
import net.dv8tion.jda.api.sharding.ShardManager;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;

import java.io.IOException;
import java.lang.reflect.Method;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.locks.ReentrantLock;

public final class ApplicationCommandsBuilder {
	private static final Logger LOGGER = Logging.getLogger();
	private final ExecutorService es = Executors.newFixedThreadPool(Math.min(4, Runtime.getRuntime().availableProcessors()));
	private final BContextImpl context;
	private final List<Long> slashGuildIds;

	private final ReentrantLock globalLock = new ReentrantLock();
	private final Map<Long, ReentrantLock> lockMap = Collections.synchronizedMap(new HashMap<>());

	public ApplicationCommandsBuilder(@NotNull BContextImpl context, List<Long> slashGuildIds) {
		this.context = context;
		this.slashGuildIds = slashGuildIds;
		this.context.setSlashCommandsBuilder(this);
	}

	public void processApplicationCommand(ApplicationCommand applicationCommand, Method method) {
		try {
			if (method.isAnnotationPresent(JDASlashCommand.class)) {
				processSlashCommand(applicationCommand, method);
			} else if (method.isAnnotationPresent(JDAUserCommand.class)) {
				processUserCommand(applicationCommand, method);
			} else if (method.isAnnotationPresent(JDAMessageCommand.class)) {
				processMessageCommand(applicationCommand, method);
			}
		} catch (Exception e) {
			throw new RuntimeException("An exception occurred while processing application command at " + Utils.formatMethodShort(method), e);
		}
	}

	private void processUserCommand(ApplicationCommand applicationCommand, Method method) {
		ReflectionUtils.checkApplicationCommandParameter(method,
				method.getAnnotation(JDAUserCommand.class).scope(),
				GlobalUserEvent.class,
				GuildUserEvent.class);

		final UserCommandInfo info = new UserCommandInfo(context, applicationCommand, method);

		final CommandPath effectivePath = context.addUserCommand(info);
		LOGGER.debug("Added user command {} for method {}", effectivePath, Utils.formatMethodShort(method));
	}

	private void processMessageCommand(ApplicationCommand applicationCommand, Method method) {
		ReflectionUtils.checkApplicationCommandParameter(method,
				method.getAnnotation(JDAMessageCommand.class).scope(),
				GlobalMessageEvent.class,
				GuildMessageEvent.class);

		final MessageCommandInfo info = new MessageCommandInfo(context, applicationCommand, method);

		final CommandPath effectivePath = context.addMessageCommand(info);
		LOGGER.debug("Added message command {} for method {}", effectivePath, Utils.formatMethodShort(method));
	}

	private void processSlashCommand(ApplicationCommand applicationCommand, Method method) {
		ReflectionUtils.checkApplicationCommandParameter(method,
				method.getAnnotation(JDASlashCommand.class).scope(),
				GlobalSlashEvent.class,
				GuildSlashEvent.class);

		final SlashCommandInfo info = new SlashCommandInfo(context, applicationCommand, method);

		final CommandPath effectivePath = context.addSlashCommand(info);
		LOGGER.debug("Added slash command path {} for method {}", effectivePath, Utils.formatMethodShort(method));
	}

	private String getCheckTypeString() {
		if (context.isOnlineAppCommandCheckEnabled()) {
			return "Online check";
		} else {
			return "Local disk check";
		}
	}

	public void postProcess() throws IOException {
		context.getJDA().setRequiredScopes("applications.commands");

		context.setApplicationCommandsCache(new ApplicationCommandsCache(context));

		scheduleGlobalApplicationCommandsUpdate(false, context.isOnlineAppCommandCheckEnabled()).whenComplete((wasUpdated, e) -> {
			if (e != null) {
				LOGGER.error("An error occurred while updating global commands", e);
			}
		});

		final Map<Guild, CompletableFuture<CommandUpdateResult>> map;
		final ShardManager shardManager = context.getJDA().getShardManager();
		if (shardManager != null) {
			map = scheduleApplicationCommandsUpdate(shardManager.getGuildCache(), false, context.isOnlineAppCommandCheckEnabled());
		} else {
			map = scheduleApplicationCommandsUpdate(context.getJDA().getGuildCache(), false, context.isOnlineAppCommandCheckEnabled());
		}

		map.forEach((guild, future) -> {
			future.whenComplete((result, throwable) -> handleApplicationUpdateException(guild, throwable));
		});
	}

	public CompletableFuture<Boolean> scheduleGlobalApplicationCommandsUpdate(boolean force, boolean online) {
		return CompletableFuture.supplyAsync(() -> {
			globalLock.lock();
			try {
				final ApplicationCommandsUpdater globalUpdater = ApplicationCommandsUpdater.ofGlobal(context, online);
				final boolean shouldUpdateCommands = force || globalUpdater.shouldUpdateCommands();
				if (shouldUpdateCommands) {
					globalUpdater.updateCommands();
					LOGGER.debug("Global commands were updated ({})", getCheckTypeString());
				} else {
					LOGGER.debug("Global commands does not have to be updated ({})", getCheckTypeString());
				}

				context.getApplicationCommandsContext().putLiveApplicationCommandsMap(null, ApplicationCommandInfoMap.fromCommandList(globalUpdater.getScopeApplicationCommands()));

				return shouldUpdateCommands;
			} catch (Throwable e) {
				throw new RuntimeException("An exception occurred while updating global commands", e);
			} finally {
				globalLock.unlock();
			}
		}, es);
	}

	void handleApplicationUpdateException(Guild guild, Throwable throwable) {
		if (throwable != null) {
			ErrorResponseException e = Utils.getErrorResponseException(throwable);

			if (e != null && e.getErrorResponse() == ErrorResponse.MISSING_ACCESS) {
				final String inviteUrl = context.getJDA().getInviteUrl() + "&guild_id=" + guild.getId();

				LOGGER.warn("Could not register guild commands for guild '{}' ({}) as it appears the OAuth2 grants misses applications.commands, you can re-invite the bot in this guild with its already existing permission with this link: {}", guild.getName(), guild.getId(), inviteUrl);
				context.getRegistrationListeners().forEach(r -> r.onGuildSlashCommandMissingAccess(guild, inviteUrl));
			} else {
				LOGGER.error("Encountered an exception while updating commands for guild '{}' ({})", guild.getName(), guild.getId(), throwable);
			}
		}
	}

	@NotNull
	public Map<Guild, CompletableFuture<CommandUpdateResult>> scheduleApplicationCommandsUpdate(@NotNull Iterable<Guild> guilds, boolean force, boolean onlineCheck) {
		final Map<Guild, CompletableFuture<CommandUpdateResult>> map = new HashMap<>();

		for (Guild guild : guilds) {
			if (!slashGuildIds.isEmpty() && !slashGuildIds.contains(guild.getIdLong())) continue;

			map.put(guild, scheduleApplicationCommandsUpdate(guild, force, onlineCheck));
		}

		return map;
	}

	@NotNull
	public CompletableFuture<CommandUpdateResult> scheduleApplicationCommandsUpdate(Guild guild, boolean force, boolean onlineCheck) {
		if (!slashGuildIds.isEmpty() && !slashGuildIds.contains(guild.getIdLong()))
			return CompletableFuture.completedFuture(new CommandUpdateResult(guild, false));

		return CompletableFuture.supplyAsync(() -> {
			final ReentrantLock lock;
			synchronized (lockMap) {
				lock = lockMap.computeIfAbsent(guild.getIdLong(), x -> new ReentrantLock());
			}

			try {
				lock.lock();

				final ApplicationCommandsUpdater updater = ApplicationCommandsUpdater.ofGuild(context, guild, onlineCheck);

				boolean updatedCommands = false;

				if (force || updater.shouldUpdateCommands()) {
					updater.updateCommands();

					updatedCommands = true;

					LOGGER.debug("Guild '{}' ({}) commands were{} updated ({})", guild.getName(), guild.getId(), force ? " force" : "", getCheckTypeString());
				} else {
					LOGGER.debug("Guild '{}' ({}) commands does not have to be updated ({})", guild.getName(), guild.getId(), getCheckTypeString());
				}

				context.getApplicationCommandsContext().putLiveApplicationCommandsMap(guild, ApplicationCommandInfoMap.fromCommandList(updater.getScopeApplicationCommands()));

				return new CommandUpdateResult(guild, updatedCommands);
			} catch (Throwable e) {
				throw new RuntimeException("An exception occurred while updating guild commands for guild '" + guild.getName() + "' (" + guild.getId() + ")", e);
			} finally {
				lock.unlock();
			}
		}, es);
	}
}

File: src/main/kotlin/com/freya02/botcommands/commands/internal/application/ApplicationCommandsBuilder.kt
package com.freya02.botcommands.commands.internal.application

import com.freya02.botcommands.api.Logging
import com.freya02.botcommands.api.annotations.Declaration
import com.freya02.botcommands.api.application.CommandUpdateResult
import com.freya02.botcommands.api.application.GlobalApplicationCommandManager
import com.freya02.botcommands.api.application.GuildApplicationCommandManager
import com.freya02.botcommands.api.application.IApplicationCommandManager
import com.freya02.botcommands.core.api.annotations.BEventListener
import com.freya02.botcommands.core.api.annotations.BService
import com.freya02.botcommands.core.internal.*
import com.freya02.botcommands.internal.BContextImpl
import com.freya02.botcommands.internal.application.ApplicationCommandInfo
import com.freya02.botcommands.internal.application.MutableApplicationCommandMap
import com.freya02.botcommands.internal.throwInternal
import com.freya02.botcommands.internal.utils.ReflectionUtilsKt.nonInstanceParameters
import com.freya02.botcommands.internal.utils.ReflectionUtilsKt.shortSignature
import com.freya02.botcommands.internal.utils.Utils
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import net.dv8tion.jda.api.JDA
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.events.guild.GuildReadyEvent
import net.dv8tion.jda.api.requests.ErrorResponse
import kotlin.reflect.full.callSuspend
import kotlin.reflect.full.valueParameters
import kotlin.reflect.jvm.jvmErasure

private val LOGGER = Logging.getLogger()

@BService
internal class ApplicationCommandsBuilder(
    private val context: BContextImpl,
    private val serviceContainer: ServiceContainer,
    classPathContainer: ClassPathContainer
) {
    private val applicationCommandsContext = context.applicationCommandsContext

    private val globalDeclarationFunctions: MutableList<ClassPathFunction> = arrayListOf()
    private val guildDeclarationFunctions: MutableList<ClassPathFunction> = arrayListOf()

    private val guildReadyMutex = Mutex()
    private val guildUpdateMutexMap: MutableMap<Long, Mutex> = hashMapOf()
    private var init = false

    init {
        for (classPathFunction in classPathContainer
            .functionsWithAnnotation<Declaration>()
            .requireNonStatic()
            .requireFirstArg(GlobalApplicationCommandManager::class, GuildApplicationCommandManager::class)
        ) {
            when (classPathFunction.function.valueParameters.first().type.jvmErasure) {
                GlobalApplicationCommandManager::class -> globalDeclarationFunctions.add(classPathFunction)
                GuildApplicationCommandManager::class -> guildDeclarationFunctions.add(classPathFunction)
                else -> throwInternal("Function first param should have been checked")
            }
        }

        LOGGER.debug("Loaded ${globalDeclarationFunctions.size} global declaration functions and ${guildDeclarationFunctions.size} guild declaration functions")
        if (globalDeclarationFunctions.isNotEmpty()) {
            LOGGER.trace("Global declaration functions:\n" + globalDeclarationFunctions.joinToString("\n") { it.function.shortSignature })
        }

        if (guildDeclarationFunctions.isNotEmpty()) {
            LOGGER.trace("Guild declaration functions:\n" + guildDeclarationFunctions.joinToString("\n") { it.function.shortSignature })
        }
    }

    @BEventListener
    internal suspend fun onGuildReady(event: GuildReadyEvent, context: BContextImpl) {
        guildReadyMutex.withLock {
            val isFirstRun = synchronized(this) {
                if (init) return@synchronized false
                init = true

                true
            }

            if (isFirstRun) {
                onFirstRun(context, event.jda)
            }
        }

        val guild = event.guild
        LOGGER.debug("Guild ready: $guild")

        try {
            updateGuildCommands(guild)
        } catch (t: Throwable) {
            handleGuildCommandUpdateException(guild, t)
        }
    }

    internal fun handleGuildCommandUpdateException(guild: Guild, t: Throwable) {
        val e = Utils.getErrorResponseException(t)

        if (e != null && e.errorResponse == ErrorResponse.MISSING_ACCESS) {
            val inviteUrl = "${guild.jda.getInviteUrl()}&guild_id=${guild.id}"
            LOGGER.warn(
                "Could not register guild commands for guild '{}' ({}) as it appears the OAuth2 grants misses applications.commands, you can re-invite the bot in this guild with its already existing permission with this link: {}",
                guild.name,
                guild.id,
                inviteUrl
            )

            //TODO missing access event
        } else {
            LOGGER.error(
                "Encountered an exception while updating commands for guild '{}' ({})",
                guild.name,
                guild.id,
                t
            )
        }
    }

    internal suspend fun updateGlobalCommands(force: Boolean = false): CommandUpdateResult {
        val manager = GlobalApplicationCommandManager(context)
        globalDeclarationFunctions.forEach { classPathFunction ->
            runDeclarationFunction(classPathFunction, serviceContainer, manager)
        }

        val globalUpdater = ApplicationCommandsUpdaterKt.ofGlobal(context, manager)
        val needsUpdate = force || globalUpdater.shouldUpdateCommands()
        if (needsUpdate) {
            globalUpdater.updateCommands()
            LOGGER.debug("Global commands were{} updated ({})", getForceString(force), getCheckTypeString())
        } else {
            LOGGER.debug("Global commands does not have to be updated ({})", getCheckTypeString())
        }

        applicationCommandsContext.putLiveApplicationCommandsMap(null, globalUpdater.guildApplicationCommands.toApplicationCommandMap())

        return CommandUpdateResult(null, needsUpdate)
    }

    internal suspend fun updateGuildCommands(guild: Guild, force: Boolean = false): CommandUpdateResult {
        guildUpdateMutexMap.computeIfAbsent(guild.idLong) { Mutex() }.withLock {
            //TODO check if guild is in slash guild IDs, if not and not empty then skip
            val manager = GuildApplicationCommandManager(context, guild)
            guildDeclarationFunctions.forEach { classPathFunction ->
                runDeclarationFunction(classPathFunction, serviceContainer, manager)
            }

            val guildUpdater = ApplicationCommandsUpdaterKt.ofGuild(context, guild, manager)
            val needsUpdate = force || guildUpdater.shouldUpdateCommands()
            if (needsUpdate) {
                guildUpdater.updateCommands()
                LOGGER.debug("Guild '${guild.name}' (${guild.id}) commands were{} updated ({})", getForceString(force), getCheckTypeString())
            } else {
                LOGGER.debug("Guild '${guild.name}' (${guild.id}) commands does not have to be updated ({})", getCheckTypeString())
            }

            applicationCommandsContext.putLiveApplicationCommandsMap(guild, guildUpdater.guildApplicationCommands.toApplicationCommandMap())

            return CommandUpdateResult(guild, needsUpdate)
        }
    }

    private fun getForceString(force: Boolean): String = if (force) " force" else ""

    private fun getCheckTypeString(): String =
        if (context.isOnlineAppCommandCheckEnabled) "Online check" else "Local disk check"

    private fun List<ApplicationCommandInfo>.toApplicationCommandMap() = MutableApplicationCommandMap.fromCommandList(this)

    private suspend fun onFirstRun(context: BContextImpl, jda: JDA) {
        LOGGER.debug("First ready") //TODO runInitialization ? (+ exit if error ?)

        jda.setRequiredScopes("applications.commands")

        context.serviceContainer.putService(ApplicationCommandsCacheKt(jda))

        try {
            updateGlobalCommands()
        } catch (e: Throwable) {
            LOGGER.error("An error occurred while updating global commands", e)
        }
    }

    private suspend fun runDeclarationFunction(
        classPathFunction: ClassPathFunction,
        serviceContainer: ServiceContainer,
        manager: IApplicationCommandManager
    ) {
        val function = classPathFunction.function
        val args = serviceContainer.getParameters(
            function.nonInstanceParameters.map { it.type.jvmErasure },
            mapOf(manager::class to manager)
        ).toTypedArray()

        function.callSuspend(classPathFunction.instance, *args)
    }
}